x/dualstaking/keeper/grpc_query_delegator_providers.go

```diff
@@ -0,0 +1,50 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/utils"
+	"github.com/lavanet/lava/x/dualstaking/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (k Keeper) DelegatorProviders(goCtx context.Context, req *types.QueryDelegatorProvidersRequest) (*types.QueryDelegatorProvidersResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	nextEpoch, err := k.getNextEpoch(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	providers, err := k.GetDelegatorProviders(ctx, req.Delegator)
+	if err != nil {
+		return nil, err
+	}
+
+	var delegations []types.Delegation
+	for _, provider := range providers {
+		indices := k.delegationFS.GetAllEntryIndicesWithPrefix(ctx, provider)
```

Comment:
this will give all delegations of the provider, not only those for the specific delegator we're interested in. so if the provider has e.g. 2 delegations, we will return both even though one does is not relevant to the query.

you need to decode the index into <provider, chain, delegator>, and pick only those with the matching delegator.

note that this will be become simpler when delegation is per provider and not per <provider,chainID> - once the provider is known we could dig up <provider,delegator> directly with FindEntry() without needing to loop over the indices.

Reply:
done. see https://github.com/lavanet/lava/commit/2f9f3f255ed7898d378b890621825f64a86bcb3b